parsers/CA_SK.py

```diff
@@ -0,0 +1,143 @@
+    This is because the API only returns daily data but the backend expects hourly values.
+    This is in order to facilitate the estimation of the hourly values from the daily average.
+    """
+    # Validate that the zone key is equal to CA-SK.
+    validate_zone_key(zone_key)
+    # Validate that the target_datetime is None as this parser is unable to fetch historical data.
+    validate_no_datetime(target_datetime, zone_key)
+
+    session = session or Session()
+
+    # Set the headers to mimic a user browser as the API will return a 403 if not.
+    headers = {"user-agent": USER_AGENT}
+    response: Response = session.get(PRODUCTION_URL, headers=headers)
+
+    if not response.ok:
+        raise ParserException(
+            "CA_SK.py",
+            f"Failed to fetch production data. Response Code: {response.status_code}\nError:\n{response.text}",
+            zone_key,
+        )
+
+    raw_data = response.json()
+    # Date is in the format "Jan 01, 2020"
+    raw_date = raw_data["SupplyDataText"]
+    date = datetime.strptime(raw_date, "%b %d, %Y")
+    production_data = {}
+
+    for value in raw_data["PowerCacheData"]["generationByType"]:
+        production_data[PRODUCTION_MAPPING[value["type"]]] = value[
+            "totalGenerationForType"
+        ]
+
+    data_list: List[dict] = []
+    # Hack to return hourly data from daily data for the backend as it expects hourly data.
+    for hour in range(0, 24):
+        data_list.append(
+            {
+                "zoneKey": zone_key,
+                "datetime": date.replace(hour=hour, tzinfo=TIMEZONE),
+                "production": production_data,
+                "source": "saskpower.com",
+            }
+        )
+
+    return data_list
+
+
+def fetch_consumption(
+    zone_key: str = "CA-SK",
+    session: Optional[Session] = None,
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+):
+    # Validate that the zone key is equal to CA-SK.
+    validate_zone_key(zone_key)
+    # Validate that the target_datetime is None as this parser is unable to fetch historical data.
+    validate_no_datetime(target_datetime, zone_key)
+
+    session = session or Session()
+
+    # Set the headers to mimic a user browser as the API will return a 403 if not.
+    headers = {"user-agent": USER_AGENT}
+
+    response: Response = session.get(CONSUMPTION_URL)  # , headers=headers)
+
+    if not response.ok:
+        raise ParserException(
+            "CA_SK.py",
+            f"Failed to fetch consumption data. Response Code: {response.status_code}\nError:\n{response.text}",
+            zone_key,
+        )
+
+    raw_data = response.json()
+
+    now = datetime.now(TIMEZONE)
+
+    # Data is updated every 5 minutes so we assume the data is from a multiple of 5 minutes and has a 5 minute delay from that multiple.
+    assumed_datetime = now.replace(second=0, microsecond=0) - timedelta(
+        minutes=(now.minute % 5) + 5
```

Comment:
```suggestion
        minutes=(now.minute % 5) * 5
```

ex: if it's 09:24 we want to say that it's valid for 09:20

Reply:
The modulus in the timedelta already does that but I also remove an additional 5 minutes because I noticed that the data never change over 5 minute periods and that if you run it at 12:00:01 for example you get the data for 12:00:00 which seems highly unlikely to be accurate as the delay between all the systems should be a few seconds at least. (Unless they are using estimates).

With the current implementation if you run it at 12:07:00 you get the data for 12:00:00 and if you run it at 12:10:00 you get the data for 12:05:00. But as I mentioned I'm not sure about this delay so I can remove if you want.

Reply:
Sorry, I misread, makes sense to introduce that delay